python/GafferUI/NodeSetEditor.py

```diff
@@ -82,15 +82,22 @@ def __init__( self, topLevelWidget, scriptNode, **kw ) :
 	# to show them. This also calls `nodeSet.setRemoveOrphans( True )` so that
 	# deleted nodes are not visible in the UI.
 	#
-	# This will break any editor links, if set.
+	# This will break any editor links, if set, unless removeDriver is False,
+	# in which case this will update the nodeSet of the driving editor instead
+	# (@see drivingEditor).
 	#
-	# The driver will be updated *after* the node set, such that calling
-	# `getNodeSetDriver` in the nodeSetChangedSignal will return the departing
-	# driver. TODO: We need to work out a sensible way to signal once state has
-	# stabilised
-	def setNodeSet( self, nodeSet ) :
+	# When cleared, the editor's driver state will be updated *after* the node
+	# set, such that calling `getNodeSetDriver` in the nodeSetChangedSignal
+	# will return the departing driver. TODO: We need to work out a sensible
+	# way to signal once state has stabilised.
+	def setNodeSet( self, nodeSet, removeDriver = True ) :
```

Comment:
I'm not convinced that this is a good API change. In the past we justified the additional complexity of drivers by saying that they were a layer on top of the (original, nice, simple) nodeSet API, but here we're muddying things by mixing the two together. I'm also unconvinced that it's clearer from a readability perspective; without it I would call `editor.drivingEditor().setNodeSet( x )`, making it absolutely clear that I'm intentionally modifying the _driver's_ state. With it, I call `editor.setNodeSet( x, removeDriver = False )`, giving the impression that I'm still operating on `editor`, which I am not.

If I'm reading correctly, it seems that the following two commits can be implemented pretty easily without this one, so I suggest we just drop it.

p.s. As an aside, one thing I like about the "Editors are Nodes" concept we've been knocking about is that we could represent the driver/driven concept nicely using connections between plugs on the editors.

Reply:
The only snag there is that `editor.drivingEditor().setNodeSet( x )` would raise a `RuntimeError` if the editor isn't linked as `drivingEditor()` returns `None`.

So the code we duplicate wherever we do this is:

```
toUpdate = editor
if editor.drivingEditor() is not None :
   toUpdate = editor.drivingEditor()
toUpdate.setNodeSet( n )
```
Which isn't the end of the world, and there are 5 different ways you can write that more concisely. I get what you mean about it looking like your modifying 'editor' but not (though ultimately are in a way)

Reply:
Yep, I knew about that, but I stand by comments. It's just a two liner instead of a one liner :

```
target = editor.driverEditor() or editor
target( n )
```

Reply:
FIxed in https://github.com/GafferHQ/gaffer/pull/3360/commits/35a7ee2b1f9e3ce2739abf73a9d015c346b698b6 and https://github.com/GafferHQ/gaffer/pull/3360/commits/d6f7e6feda28a9f1f3c5a76dc33caa0de8102d7d